### PR TITLE
Implement guest-agent-exec probes

### PR DESCRIPTION
**Pull request by kwiesmueller**
**What this PR does / why we need it**:
Add Exec to VM/VMI livenessProbe and readinessProbe so users can run commands for determining VM health.
The commands supplied to those probes will be wrapped by a binary (`virt-probe`) inside the virt-launcher container and forwarded to the qemu-guest-agent if present. 

**Special notes for your reviewer**:
The changes are still a little messy. I'm not entirely happy with the command and result passing around between launcher, server, cli etc.
Took me a bit to dig into, happy for some insight and suggestions.

Refer to [probes.md](https://github.com/kwiesmueller/kubevirt/blob/qemu-exec-probes/docs/probes.md) for an example.

**Release note**:
```release-note
Add exec command to VM liveness and readinessProbe executed through the qemu-guest-agent.
```

/cc @fabiand @omeryahud @davidvossel @rmohr

### CODE REVIEW

File: docs/probes.md
Comment (davidvossel) [inline docs/probes.md:8]: worth noting in this documentation that this probe behavior requires not only the guest agent to be installed, but also for the guest-exec api to be enabled within the guest.

For security reasons, this guest agent command is often disabled by default. For example, rhel8 disables it by default

File: docs/probes.md
Comment (davidvossel) [inline docs/probes.md:50]: as an example, i'd give this a slightly higher timeout. maybe 5 seconds? people will copy and paste this

File: pkg/virt-controller/services/template.go
Comment (davidvossel) [inline pkg/virt-controller/services/template.go:1611]: do we know how much memory this process uses? I think we'll need to add memory overhead for this execution

**Review by davidvossel — COMMENTED** [review]
Hey, nice feature. I did a high level overview and things look pretty good. I'd like to see a functional test to exercise this.

File: docs/probes.md
Comment (kwiesmueller) [reply docs/probes.md:8]: Right, I'm still investigating if we can provide images (like the new containerDisks) with the agent installed and enabled by default or at least get it there for some dists frequently used (like Fedora Cloud).

File: pkg/virt-controller/services/template.go
Comment (kwiesmueller) [reply pkg/virt-controller/services/template.go:1611]: I haven't checked, but would expect it to be very small as it's just a single gRPC call per probe run.
Don't know how tight the resource constraints are and how much is allocated for the container compared to the VM.

File: docs/probes.md
Comment (kwiesmueller) [reply docs/probes.md:50]: We should make sure the probe is safe.
Right now before v1.20 Kubernetes the timeoutSeconds would not get used correctly.
https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes

The note on that link also plays into your question regarding memory overhead as in the worst case it might cause the problem mentioned there:
```
Note:
Before Kubernetes 1.20, the field timeoutSeconds was not respected for exec probes: probes continued running indefinitely, even past their configured deadline, until a result was returned.

This defect was corrected in Kubernetes v1.20. You may have been relying on the previous behavior, even without realizing it, as the default timeout is 1 second. As a cluster administrator, you can disable the feature gate ExecProbeTimeout (set it to false) on each kubelet to restore the behavior from older versions, then remove that override once all the exec probes in the cluster have a timeoutSeconds value set.
If you have pods that are impacted from the default 1 second timeout, you should update their probe timeout so that you're ready for the eventual removal of that feature gate.

With the fix of the defect, for exec probes, on Kubernetes 1.20+ with the dockershim container runtime, the process inside the container may keep running even after probe returned failure because of the timeout.

Caution: Incorrect implementation of readiness probes may result in an ever growing number of processes in the container, and resource starvation if this is left unchecked.

```

File: pkg/virt-controller/services/template.go
Comment (kwiesmueller) [reply pkg/virt-controller/services/template.go:1611]: @davidvossel we could also use conditions instead. Then our controlplane could talk to the respective virt-launchers and run the probes from the outside and update the custom readiness condition:
https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-gate

Comment (davidvossel) [reply pkg/virt-controller/services/template.go:1611]: for now, i think this is as simple as updating the `getMemoryOverhead()` function in template.go. For every readiness/liveliness probe that calls this virt-probe wrap, we need to add some overhead.

File: docs/probes.md
Comment (davidvossel) [reply docs/probes.md:50]: ha, that would be bad.

My concern about memory is due to us getting OOM killed due to components within the pod tipping the pod over the requested/limited memory requirements.  we'lve had issues wiht that in the past, which is why we calculated memory overhead of our components in virt-launcher pod

Comment (kwiesmueller) [reply docs/probes.md:50]: I'm building a small fake cmdserver right now so I can profile the virt-probe command to see how it does memory wise.

File: docs/probes.md
Comment (davidvossel) [reply docs/probes.md:8]: i don't know how feasible that is.  the agent maintainers are pushing out builds using rpms for respective distros like fedora, rhel, suse, etc... If we come up wiht our own delivery mechanism, we have to continually build and push out new containers with the agent in it.  then we'd have to match those builds with OS versions. for example, an agent built for the new fedora 34 won't work wiht fedora 32 due to glibc requirements. 

i think the simpliest thing would be to document this requires qemu-exec to be enabled and provide an example of how to do that.

File: tests/probe_test.go
Comment (kwiesmueller) [inline tests/probe_test.go:85]: not sure if this is a safe test. it could end up to be flaky :-/
we also might not really need it with proper unit test coverage throughout the stack as it basically tests Kubernetes functionality.

**Comment by kwiesmueller** [comment]
I don't understand why the migration tests should fail because of this. They run fine locally.

/retest

**Comment by kwiesmueller** [comment]
Alright, I added more tests and also some functests.
This should now be ready for a proper review.

File: tests/credentials_test.go
Comment (davidvossel) [inline tests/credentials_test.go:41]: is this being used outside of the credentials_test? if so lets move it to some shared package

File: cmd/virt-probe/virt-probe.go
Comment (davidvossel) [inline cmd/virt-probe/virt-probe.go:48]: what is the advantage of having virt-probe talk to virt-launcher which then talks connectso to the guest agent? Could we not get the same result wiht having virt-probe execute this command with the client itself?

**Review by davidvossel — COMMENTED** [review]
looks like the right direction. I had one comment about the usefulness of exposing a new api interface for the exec command for virt-launcher.

File: cmd/virt-probe/virt-probe.go
Comment (kwiesmueller) [reply cmd/virt-probe/virt-probe.go:48]: It hands of the logic on how to talk to the guest directly (which socket etc.) to virt-launcher.
I didn't want to duplicate the logic and also wanted to keep virt-probe as small as possible.
Also we thought about exposing the exec functionality as an API at some point so it would help to already have it in the api.

File: tests/credentials_test.go
Comment (kwiesmueller) [reply tests/credentials_test.go:41]: I use it in the probes_test as well now, but not sure if I can get rid of it there.
Could be worth moving.

Comment (davidvossel) [reply tests/credentials_test.go:41]: lets move it to tests/utils.go at least since it is shared.

**Review by davidvossel — COMMENTED** [review]
/approve

i;d just like to see the shared func test function moved to a library before we merge

File: tests/utils.go
Comment (davidvossel) [inline tests/utils.go:4906]: any reason why you're using the RestClient here instead of the VirtualMachineInstance ?
```
vmi, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmi)
```

**edit** leaving this because it's funny. You did this because you moved code I wrote, and i have no idea why I wrote it this way, lol.

File: tests/utils.go
Comment (davidvossel) [inline tests/utils.go:4914]: I just noticed we're ignoring warnings here. There was a specific reason for this in the credentials_test that is known, and the reason exists for your probe_test as well.  It doesn't make sense generically for all tests though.

Maybe rename VMILauncherIgnoreWarnings ? or something like that just so it's obvious?

**Comment by kwiesmueller** [comment]
Switched it to a draft to no longer trigger tests on it until CI is happy with us again.

**Comment by rmohr** [comment]
> Add Exec to VM/VMI livenessProbe and readinessProbe so users can run commands for determining VM health.
> The commands supplied to those probes will be wrapped by a binary (`virt-probe`) inside the virt-launcher container and forwarded to the qemu-guest-agent if present.

I think before we add more probes, we should rework the whole readiness/liveness probes. In general based on past experiences I would recommend to avoid defining exec probes as such (they are extremely unpredictable regarding tho requests and limits  on pods, we had to solve a lot of issues and at the end  removed the usage of `exec` probes as a whole from the project).

Instead, that may sound weird at first, the path forward from my point of view is to only use livenes and readiness network probes on the `virt-launcher` binary and let the `virt-launcher` binary do the real liveness and readiness probes as specified on the VMI. This way we can finally also integrate considerations regarding to paused VMIs and such, which is impossible with the current piggybacking on the pod probes.

**Comment by kwiesmueller** [comment]
We already discussed the guest-agent-ping probe which we would add a custom probe type to liveness/readiness.
Do I understand correctly, that you'd want to add endpoints to virt-launcher that can be called to trigger an exec in the guest?
So virt-launcher does the probing?
I think I suggested something similar before where we don't use the pod probes at all but use https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-gate and do the probing in one of our components. This way we'd also support multiple readiness conditions.

At the end there was a discussion and we decided that this approach is simpler and clearer as we'd have to add our own probing logic which also is more work on our controlplane.

**Comment by rmohr** [comment]
> We already discussed the guest-agent-ping probe which we would add a custom probe type to liveness/readiness.
> Do I understand correctly, that you'd want to add endpoints to virt-launcher that can be called to trigger an exec in the guest?
> So virt-launcher does the probing?
> I think I suggested something similar before where we don't use the pod probes at all but use https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-gate and do the probing in one of our components. This way we'd also support multiple readiness conditions.

I really don't like on them that this requires PATCH through the API.

> 
> At the end there was a discussion and we decided that this approach is simpler and clearer as we'd have to add our own probing logic which also is more work on our controlplane.

Our current approach makes a clean integration in pod readiness/liveness for some of our states basically impossible. I know that this is more work, but that should not be the deciding factor. We can not come to a clean integration if we don't really wrap the probes. Further I can only suggest again to avoid using `exec` probes as kubevirt-infra-part. For that I have to put a hold alone :/

**Comment by rmohr** [comment]
/hold

We can not use `exec` probes for infra workflows. Guest-agent-exec probes have to work without `exec` probes on the pod.

**Comment by davidvossel** [comment]
> We can not use exec probes for infra workflows. Guest-agent-exec probes have to work without exec probes on the pod.

Ugh... i totally forgot about the resource requests/limits issue

**Comment by rmohr** [comment]
/unhold

After offline discussions, we agreed to proceed with the exec probes. For now we add a significant amount of overhead to avoid any issues based on cpu cores (something like 100Mi).

**Comment by vladikr** [comment]
I'm wondering why do we have to execute on the guest os?
Why couldn't we simply get away with `guest-ping` ?
Also, why couldn't we just rely on something like a watchdog device?

**Comment by davidvossel** [comment]
> I'm wondering why do we have to execute on the guest os?
> Why couldn't we simply get away with guest-ping ?

the probes coordinate with k8s services. For example,  when a vmi's readiness probe succeeds, then the service adds the vmi pod into rotation for a load balancer. By having this qemu exec probe, someone can time having the vmi's pod become eligible for traffic once the applications within the guest have completely booted.

I think Kevin is following up with an explicit `guest-agent-ping` probe which will use the bones of what he's developed here in this PR. the guest agent ping will help probably for the 90% use case where people just want the VMI to be marked as ready once the OS has booted (which guest ping is probably enough).  The remaining 10% use case where people need custom logic to indicate readiness of the guest is handled by this pr... so maybe this was done backwards, but i think both are ultimately useful in the end. 

> Also, why couldn't we just rely on something like a watchdog device?

for the purposes of just ensuring the guest is active, the watchdog makes a lot of sense. probes offer more than that though.

**Comment by kwiesmueller** [comment]
/retest
The unit test failing only happens in CI for me. Trying to see if it's reproducible somehow.

**Comment by kwiesmueller** [comment]
@davidvossel @rmohr PTAL

File: pkg/virt-handler/cmd-client/client.go
Comment (davidvossel) [inline pkg/virt-handler/cmd-client/client.go:645]: using the hard coded `shortTimeout` here makes an assumption the the exec probe is something that will return within a small timeframe `5 seconds`

That's probably the case the majority of the time, but I think it's technically possible for someone to configure an exec probe with a custom timeout which might exceed this pre-defined 5 second timeout. In that case, we'd always fail.

since the timeout is part of the api, we should probably extend it to be accounted for during the execution

File: pkg/virt-controller/services/template.go
Comment (davidvossel) [inline pkg/virt-controller/services/template.go:1746]: pass the exec timeout here.

**Review by davidvossel — COMMENTED** [review]
looks good. before we merge i just wanted clarity on how to handle exec timeouts. I made some suggestions

File: pkg/virt-handler/cmd-client/client.go
Comment (kwiesmueller) [reply pkg/virt-handler/cmd-client/client.go:645]: Right, I just read through the note here again: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes
We should support the timeout specified on the probe and do the same.
It may still result in resource leakage if they user command does not terminate properly, but at least we relieve K8s.
I'll have a look.

Comment (kwiesmueller) [reply pkg/virt-handler/cmd-client/client.go:645]: Right now we also only wait for the qemu-exec command to return in 10s.
This should probably use the same value?
https://github.com/kubevirt/kubevirt/pull/5038/files#diff-a70c73881228b460a83e4bbfa48cffe8c55af0ba91fea5f48dabd3390ffdea5bR67

**Comment by kwiesmueller** [comment]
@davidvossel I added support for the timeout field and added a bit more documentation on it regarding exec probes.
Right now we don't kill the guest command as we might not know how (Windows, Linux, etc.) and it's up to the user to choose a command that does not block infinitely, but aside from that the timeout works and should avoid resource leakage because of it.

**Review by davidvossel — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
